File: backend/src/main/java/com/stokuj/books/model/entity/Book.java
package com.stokuj.books.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@Entity
@Table(name = "books")
public class Book {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String title;
    private String author;
    private int year;
    private String isbn;
    private String description;
    private int pageCount;

    // === KATEGORYZACJA ===
    @ElementCollection
    @CollectionTable(name = "book_genres", joinColumns = @JoinColumn(name = "book_id"))
    @Column(name = "genre")
    private Set<String> genres;

    @ElementCollection
    @CollectionTable(name = "book_tags", joinColumns = @JoinColumn(name = "book_id"))
    @Column(name = "tag")
    private List<String> tags;

    @OneToMany(mappedBy = "book", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("chapterNumber ASC")
    @JsonIgnore
    private List<BookChapter> chapters = new ArrayList<>();

    private int chaptersCount;
    private int nerCompletedCount;

    // === OCENY ===
    private double rating;
    private int ratingsCount;

    // === METADANE ===
    private LocalDate createdAt;
    private LocalDate updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDate.now();
        this.updatedAt = LocalDate.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDate.now();
    }
}

File: svelte-frontend/src/lib/types/book.ts
export type BookId = number;

export interface SerieInfo {
	id: number;
	name: string;
	description: string;
	status: string;
	created_at: string;
}

/** Shape returned by GET /api/books/ (BookListSerializer). */
export interface BookListItem {
	id: number;
	slug: string;
	title: string;
	authors: string[];
	cover_url: string | null;
	year: number | null;
	genres: string[];
	avg_rating: number;
}

/** Shape returned by GET /api/books/{slug}/ (BookDetailSerializer — no id). */
export interface BookDetail {
	slug: string;
	title: string;
	authors: string[];
	cover_url: string | null;
	year: number | null;
	isbn: string | null;
	genres: string[];
	tags: string[];
	description: string | null;
	page_count: number | null;
	avg_rating: number;
	ratings_count: number;
	position_in_series: number | null;
	serie: SerieInfo | null;
	created_at: string;
	updated_at: string;
}
